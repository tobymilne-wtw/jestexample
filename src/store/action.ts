import { createAction, props } from '@ngrx/store';
import { Todo } from './todo.model';

const loadTodos = createAction('[Todo] Load Todos');
const loadTodosSuccess = createAction('[Todo] Load Todos Success', props<{ todos: Todo[]; }>());
const loadTodosFailure = createAction('[Todo] Load Todos Failure', props<{ error: string; }>());
const addTodo = createAction('[Todo] Add Todo', props<{ todo: Todo; }>());
const updateTodo = createAction('[Todo] Update Todo', props<{ todo: Todo; }>());
const deleteTodo = createAction('[Todo] Delete Todo', props<{ id: string; }>());

export const TodoActions = {    
    loadTodos,
    loadTodosSuccess,
    loadTodosFailure,
    addTodo,
    updateTodo,
    deleteTodo
};
