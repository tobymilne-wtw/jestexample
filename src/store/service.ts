import { Injectable } from '@angular/core';
import { Observable, delay, of } from 'rxjs';
import { Todo } from './todo.model';

@Injectable()
export class ToDoService {
 // fake backend
  getAll(): Observable<Todo[]> {
    return of(
      [{
        id: 1,
        description: 'description 1',
        completed: false
      },
      {
        id: 2,
        description: 'description 2',
        completed: false
      }]
    ).pipe(delay(2000))
  }
}
