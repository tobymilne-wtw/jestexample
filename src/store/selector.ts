import {  createSelector } from "@ngrx/store"
import { TodoState } from "./reducer";
import { AppState } from "./store";
 

const feature = (state: AppState) => state.todo;

export const todoSelector = createSelector(
  feature,
  (state: TodoState) => state.todos
);