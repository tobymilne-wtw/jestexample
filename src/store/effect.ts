import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { TodoActions } from './action';
import { catchError, map, mergeMap, of } from 'rxjs';
import { ToDoService } from './service';

@Injectable()
export class TodoEffects {

  loadTodos$ = createEffect(() =>
    this.actions$.pipe(
      ofType(TodoActions.loadTodos),
      mergeMap(() =>
        this.todoService.getAll().pipe(
          map((todos) => TodoActions.loadTodosSuccess({ todos })),
          catchError((error) =>
            of(TodoActions.loadTodosFailure({ error: error.message }))
          )
        )
      )
    )
  );
  constructor(private actions$: Actions, private todoService: ToDoService) {}
}
