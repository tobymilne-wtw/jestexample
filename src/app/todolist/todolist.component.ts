import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Signal, OnInit } from '@angular/core';
import { Todo } from '../../store/todo.model';
import { AppState } from '../../store/store';
import { Store } from '@ngrx/store';
import { TodoActions } from '../../store/action';
import { todoSelector } from '../../store/selector';
import { toSignal } from '@angular/core/rxjs-interop'

@Component({
  selector: 'app-todolist',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './todolist.component.html',
  styleUrl: './todolist.component.css',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TodolistComponent implements OnInit {
  todos$: Signal<Todo[] | undefined>;
  isLoading$: Signal<boolean | undefined>;

  constructor(public store: Store<AppState>) {
    this.todos$ = toSignal(this.store.select(todoSelector)); // observable converted to signal
    this.isLoading$ = toSignal(this.store.select((state) => state.todo.loading));
  }

  ngOnInit() {
   this.loadTodos();
  }

  loadTodos() {
    this.store.dispatch(TodoActions.loadTodos());
  }

  addTodo(index: number) {
    const todo: Todo = { id: index, description: `description ${index}`, completed: false };
    this.store.dispatch(TodoActions.addTodo({ todo }));
  }

  complete(todo: Todo) {
    this.store.dispatch(
      TodoActions.updateTodo({ todo: { ...todo, completed: true } })
    );
  }
 }


