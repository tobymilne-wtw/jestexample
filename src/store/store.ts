import { Action, ActionReducer } from '@ngrx/store';
import { TodoState, todoReducer } from './reducer';
import { TodoEffects } from './effect';

export interface AppState {
    todo: TodoState
  }
  
  export interface AppStore {
    todo: ActionReducer<TodoState, Action>;
  }
  
  export const appStore: AppStore = {
    todo: todoReducer
  }
  
  export const appEffects = [TodoEffects];
  