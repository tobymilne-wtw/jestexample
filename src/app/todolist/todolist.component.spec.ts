import { TestBed } from '@angular/core/testing';
import { provideMockStore } from '@ngrx/store/testing';
import { TodolistComponent } from './todolist.component';
import { AppState } from '../../store/store';
import { TodoActions } from '../../store/action';


const initialState: AppState = {todo: {todos:[
    { id: 1, description: 'description 1', completed: false },
], loading: false, error: 'Some error'}};

describe('TodoListComponent', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [TodolistComponent],
      providers: [provideMockStore({initialState})]
    }).compileComponents();
  });

  test('create the component', () => {
    const fixture = TestBed.createComponent(TodolistComponent);
    const app = fixture.componentInstance;
    expect(app).toBeTruthy();
  });

  test('render correctly', () => {
    const fixture = TestBed.createComponent(TodolistComponent);
    fixture.detectChanges();
   
    expect(fixture).toMatchSnapshot();
  });

  test('dispatch load action on init', () => {

    const action = TodoActions.loadTodos();

    const fixture = TestBed.createComponent(TodolistComponent);

    const dispatchSpy = jest.spyOn(fixture.componentInstance.store, 'dispatch');

    fixture.detectChanges();
   
    expect(dispatchSpy).toHaveBeenCalledWith(action);
  });

  test('dispatch add action on add', () => {

    const action = TodoActions.addTodo({ todo: { id: 1, description: 'description 1', completed: false } });

    const fixture = TestBed.createComponent(TodolistComponent);

    const dispatchSpy = jest.spyOn(fixture.componentInstance.store, 'dispatch');

    fixture.detectChanges();
    fixture.componentInstance.addTodo(1);
   
    expect(dispatchSpy).toHaveBeenCalledWith(action);
  });
});