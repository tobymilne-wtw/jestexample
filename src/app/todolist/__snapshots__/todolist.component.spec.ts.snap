// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListComponent render correctly 1`] = `
<app-todolist
  isLoading$={[Function Function]}
  store={[Function _MockStore]}
  todos$={[Function Function]}
>
  <ul>
    <li>
       1 description 1 - Pending 
      <button>
         complete 
      </button>
    </li>
  </ul><button>
    Refresh
  </button><button>
    Add
  </button>
</app-todolist>
`;
